README.md
PVSio-web [![Build Status](https://travis-ci.org/thehogfather/pvsio-web.svg?branch=alpha)](https://travis-ci.org/thehogfather/pvsio-web) [![Latest Stable Version](https://img.shields.io/npm/v/pvsio-web.svg?style=flat-square)](https://www.npmjs.com/package/pvsio-web) [![NPM Downloads](https://img.shields.io/npm/dt/pvsio-web.svg?style=flat-square)](https://www.npmjs.com/package/pvsio-web)
=========

PVSio-web is a new graphical tool to prototype and analyse user interface software. It provides the typical functionalities of an animation environment that allows designers to load and simulate realistic user interfaces. Underneath, it uses SRI's state-of-the-art theorem prover PVS for analysis, and the pvsio component as a basis for simulation.

Simulations created with PVSio-web can be watched in this youtube video https://www.youtube.com/watch?v=T0QmUe0bwL8

Live version
------------

Realistic prototypes created using PVSio-web can be found at the following links:
* http://www.pvsioweb.org/demos/Radical7 (FDA's Generic Infusion Pump prototype - full model)
* http://www.pvsioweb.org/demos/GPCA     (Commercial patient monitor - monitoring mode only)
* http://www.pvsioweb.org/demos/AlarisGP (Commercial infusion pump prototype - full model)
* http://www.pvsioweb.org/demos/BBraun   (Commercial infusion pump prototype - data entry system only)

The full PVSio-web tool with limited features is also available at http://www.pvsioweb.org
(please note that the web server has limited processing power so its response time might not be optimal; also, note that the file system is read-only, so you will not be able to save new prototypes or compile new models using this live version).

![Screenshot](screenshot.png?raw=true)


Installation
------------
To install PVSio-web, first you need to install PVS and NodeJS, and then clone the PVSio-web github repository. This can be done as follows.

#### Step 1: Install PVS and add PVS executables to your PATH
PVS is required to run PVSio-web. The tool can be downloaded from http://pvs.csl.sri.com/download.shtml Installation instructions are on the aforementioned website.

Once PVS is installed, please add the following PVS executable files to your PATH: pvs, pvsio and proveit. A simple way to do this is to create symbolic links to those files, and place the symbolic links in /usr/bin. For instance, if PVS is installed in /opt/PVS/pvs, the following commands executed in a Terminal window create the required symbolic links:

    sudo ln -s /opt/PVS/pvs /usr/bin/pvs
    sudo ln -s /opt/PVS/pvsio /usr/bin/pvsio
    sudo ln -s /opt/PVS/proveit /usr/bin/proveit

Please note that the ln command requires a full path.

#### Step 2: Install NodeJS
NodeJS is required to run PVSio-web. Please download and install NodeJS from http://nodejs.org
Installation instructions are on the aforementioned website.

#### Step 3: Clone the PVSio-web repository
Create a directory where you would like to install PVSio-web on your local computer. Open a Terminal window in the created directory, and use 'git' to clone the PVSio-web repository:

    git clone https://github.com/thehogfather/pvsio-web.git
    cd pvsio-web
    npm install

PVSio-web is now installed on your local computer!

Running pvsio-web
-----------------
To run pvsio-web, a backend and a frontend need to be started.

To start the backend: open a Terminal window in the pvsio-web directory, and use the following command (and leave the Terminal window open):

    ./start.sh

To start the frontend: open a browser (Firefox 21 or greater, or Chrome), and type the following address in the address bar:

    http://localhost:8082/

Updating pvsio-web
------------------
To update pvsio-web to the latest version, open a Terminal window, and execute the following command from the pvsio-web directory:

    git pull


Examples
--------
Realistic simulations created with PVSio-web can be watched in this youtube video:
* https://www.youtube.com/watch?v=T0QmUe0bwL8

All simulation examples demonstrated in the youtube video are included in the PVSio-web distribution in examples/projects. To open these examples, start pvsio-web and click the "Open Projects" button of the pvsio-web frontend and select one of the examples from the list.


Directory structure
-------------------
This project has the following setup:

* start.sh - the script used to initiate the server.
* examples/ - this directory contains projects and demos
* src/ - this directory contains the pvsio-web source code
    * client/ - this directory contains the source code for the pvsio-web client. This code is executed in the user's browser
    * server/ - this directory contains the source code for the pvsio-web server. This code is executed in the node.JS environment, and manages communication between pvs/pvsio and the client code.


Nightly builds
--------------
To obtain the latest development versions of pvsio-web, you can clone our alpha branch on the github repository.

To clone the alpha branch, create a new directory (for example, pvsioweb-alpha), open a Terminal window, and execute the following commands from the created directory:

    git clone https://github.com/thehogfather/pvsio-web.git -b alpha
    cd pvsio-web
    npm install


Testing the installation
------------------------
To test the client, start the pvsio-web backend by running the following command in a Terminal window (and leave the Terminal window open):
    ./start.sh

and type the following address in a browser window:

    http://localhost:8082/tests

To test the server, run the following command in a Terminal window

    npm test


Uninstallation :(
--------------
To uninstall, delete the pvsio-web folder from your computer.


CHANGELOG.md
Changelog
==========

3.0.0 (Mar 2021)
---------------
    Porting to typescript
    Revised architecture
    New version of prototype builder

2.3.0 (Dec 2017 / Jan 2018)
---------------
    Added tutorials for PVSio-web APIs (see http://www.pvsioweb.org/tutorials)
    Added new widgets libraries: core, media
    Added new demo: Medicaltech MT32-evo
    Added new code generator: JavaScript
    Added new Playback Player APIs for creating training material
    Added online documentation for widgets APIs (see http://www.pvsioweb.org/docs/widgets-library)

2.2.0 (June 2017)
---------------
    Added new demo: Medrad Stellant CT Injection System
    Added new demo: Pillbox device prototype
    Added new prototype: Niki T34 syringe driver
    Added new widget: Syringe, featuring: moving plunger, fittings (needle, spiral infusion set, vial), two size (20mL, 200mL)
    Added proof strategies for Emucharts verification templates
    Improved Emucharts Editor user interface

2.1.1 (Dec 2016)
---------------
    Added new feature: PVS verification templates for Emucharts
    Added new feature: UPPAAL file importer for Emucharts
    Improved overall PVSio-web user interface

2.1 (Sep/Oct 2016)
---------------
    Added new PVSio-web tool: Storyboard editor/simulator
    Added new widgets: Touchscreen elements
    Added new example project: Flight Control Unit
    Improved Prototype Builder (live widgets preview, direct support for different types of buttons, displays, LEDs)
    Refactored widgets library

2.0.2 (April 2016)
---------------
    PVSio-web goes live on UMinho servers (http://pvsioweb.lsd.di.uminho.pt)
    Added new feature: physical keyboard keys can be linked to prototype keys
    Added new feature: voice readback for button widgets
    Added new feature: wallclock for automatic execution of tick transitions defined in the PVS model
    Improved Emucharts Editor

2.0.1b (Sep 2015)
---------------
    Added new modelling extension: Presentation Interaction Models (PIMs)
    Added new feature: test case generation for PIM / Emucharts
    Added PVSio-web network controller for distributed simulations

2.0.1 (April 2015)
---------------
    Added new Model Generators: VMD-SL (Vienna Development Method)
    Added new prototypes (full Alaris GP, Pacemaker, GPCA-Simulink)
    Added new display widgets (multi-line displays, LEDs)
    Added new file browser

2.0.0 (Jan 2015)
---------------
    PVSio-web goes live on the Heroku cloud (http://www.pvsioweb.org)

1.1.0 (Dec 2014)
---------------
    Added new Model Generators: MAL (Modal Action Logic), and PIM (Presentation Interaction Models)
    Improved Prototype Builder (widgets selection list, image scale)
    Improved server backend (file system watcher, autosave current project)

1.0.0 (July 2014)
---------------
    Added changelog file
    Added grunt file configurations for building project
    Added new protytpes (Baxter Sigma, Smiths Medical MedFusion, Alaris PC)
    Added new tool: Emucharts Editor (replaces Emulink, new architecture/implementation, supports multiple model/code generators)
    Added new tool: Model Editor (replaces Text Editor, new implementation based on CodeMirror)
    Improved tool architecture and code quality (JSHint static analyser, Jasmine unit testing)
    PVSio-web goes live on the Amazon cloud

0.0.18 (Mar 2014)
---------------
    Added support for logging/playback of user interactions
    Added voice feedback function for display widgets
    Added new protytpes (Arcomedical Syramed)

0.0.17 (Dec 2013)
---------------
    Added new tool: Emulink, for creating pvs models using Statechart-like graphical diagrams
    PVSio-web license changes from MIT to GPL3

0.0.16 (Nov 2013)
---------------
    Added new prototypes (BBraun Perfusor, Baxter Colleague, Zymed Syringe, TCAS)
    Improved Prototype Builder (cursor displays)

0.0.15 (Sep 2013)
---------------
    Added new tool: PVSio-web browser for navigating the file system
    Added logging mechanisms
    Improved Text Editor (autocomplete, syntax highlighting)
    Improved prototypes (full GPCA demo based on the FDA Simulink Model)

0.0.13 (Jun 2013)
---------------
    Improved architecture (introduced callback mechanisms for sending/receiving pvsio commands)

0.0.10 (Apr 2013)
---------------
    Added new tool: Text Editor for editing PVS models
    Introduced templating mechanism based on underscore.js

0.0.6 (Feb 2013)
---------------
    Added example prototypes (Alaris GP, Alaris Asena, GPCA)
    Improved Prototype Builder (save/load projects, widgets creation wizard)
    Improved server backend to support multiple clients

0.0.5 (Dec 2012)
---------------
    First public release of PVSio-web, featuring a Prototype Builder for developing interactive prototypes based on formal models


pvssocketserver.js
//#!/usr/bin/env node
/**
Copyright (c) 2012

Patrick Oladimeji
This file is part of pvsio-web.

pvsio-web is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

pvsio-web is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with Foobar.  If not, see <http://www.gnu.org/licenses/>.
*/
/**
 * This file creates a connection to a pvsio process run locally or at specified host.
 * It also creates an express webserver to serve demo applications e.g. the infusion
 * pump pvsio demo.
 * The websocket connection started by this process listens for 3 commands:
 * sendCommand: used to send a pvsio command to the processs
 * startProcess: used to start the pvsio process
 * getSourceCode: used to get the source code of the pvs code being executed
 * @author patrick
 * @date 28 Jul 2012 21:52:31
 *
 */
/*jshint unused: true, undef: true*/
/*jslint vars: true, plusplus: true, devel: true, nomen: true, indent: 4, maxerr: 50, undef: true, node: true */
/*global __dirname*/

function run() {
    "use strict";
    
    var pvsio                   = require("./pvsprocess"),
        path                    = require("path"),
        ws                      = require("ws"),
        http                    = require("http"),
        fs                      = require("fs"),
        express                 = require("express"),
        bodyParser              = require("body-parser"),
        webserver               = express(),
        procWrapper             = require("./processwrapper"),
        uploadDir               = "/public/uploads",
        port                    = 8082,
        pvsioProcessMap         = {},//each client should get his own process
        httpServer              = http.createServer(webserver),
        Promise                 = require("es6-promise").Promise,
        logger                  = require("tracer").console(),
		serverFuncs				= require("./serverFunctions"),
        baseProjectDir          = __dirname + "/public/projects/";
    var p, clientid = 0, WebSocketServer = ws.Server;
	var writeFile = serverFuncs.writeFile,
		createProject = serverFuncs.createProject,
		openProject = serverFuncs.openProject,
		listProjects = serverFuncs.listProjects;

    /**
     * Utility function that dispatches responses to websocket clients
     * @param {{type:string, data}} token The token to send to the client
     * @param {Socket} socket The websocket to use to send the token
     */
    function processCallback(tok, socket) {
        //called when any data is recieved from pvs process
        //if the type of the token is 'processExited' then send message to client if the socket is still open
        if (tok.type === 'processExited') {
            if (socket.readyState === 1) {
                socket.send(JSON.stringify(tok));
            }
        } else {//send the message normally
            socket.send(JSON.stringify(tok));
        }
    }

    /**
     * reads and changes the settings in the .pvsioweb file in the project root
     * @param {string} projectName the name of the project
     * @param {string} key the key of the setting to write
     * @param {object} value the value of the setting to write
     * @returns {Promise} a promise that is resolved when the settings file has been written.
    */
    function changeProjectSetting(projectName, key, value) {
        var file = path.join(baseProjectDir, projectName, "/.pvsioweb"),
            props = {};
        return new Promise(function (resolve, reject) {
            //if file does not exist, create it. Else read the property file and update just the key value specified
            fs.exists(file, function (exists) {
                if (!exists) {
                    props[key] = value;
                    writeFile(file, JSON.stringify(props, null, " "))
                        .then(resolve, reject);
                } else {
                    fs.readFile(file, {encoding: "utf8"}, function (err, res) {
                        props = {err: err};
                        if (!err) {
                            props =  JSON.parse(res) || props;
                            props[key] = value;
                            //write the file back
                            writeFile(file, JSON.stringify(props, null, " "))
                                .then(resolve, reject);
                        } else {//there was an error so reject the promise
                            reject(err);
                        }
                    });
                }
            });
        });
        
    }
    
    //create logger
    webserver.use("/demos", function (req, res, next) {
        logger.log('Method: %s,  Url: %s, IP: %s', req.method, req.url, req.connection.remoteAddress);
        next();
    });
    //create the express static server and use public dir as the default serving directory
    webserver.use(express.static(__dirname + "/public"));
    webserver.use(bodyParser({ keepExtensions: true, uploadDir: __dirname + uploadDir}));

    /**
     * used to manage file upload process for pvsio-web
     */
    webserver.all("/upload", function (req, res) {
        logger.debug(JSON.stringify(req.files));
        var fileName = req.files.file.path.split("/").slice(-1).join("");
        //should return a map of oldname to new name for the uploaded files
        res.send({fileName: fileName});
    });

    function typeCheck(filePath, cb) {
        procWrapper().exec({
            command: "proveit " + filePath,
            callBack: cb
        });
    }
    
    /**
        get function maps for client sockets
    */
    function createClientFunctionMaps() {
        var map = {
            "renameFile": function (token, socket, socketid) {
                var oldPath = path.join(baseProjectDir, token.oldPath);
                var newPath = path.join(baseProjectDir, token.newPath);
                fs.rename(oldPath, newPath, function (err) {
                    if (err) {
                        logger.debug("warning, error while renaming " + token.oldPath +
                                    " into " + token.newPath + " (" + err + ")");
                    }
                    processCallback({id: token.id, socketId: socketid, err: err}, socket);
                });
            },
            "readDirectory": function (token, socket, socketid) {
                fs.readdir(token.path, function (err, files) {
                    if (!err) {
                        //get stat attributes for all the files using an async call
                        var promises = files.map(function (f) {
                            return new Promise(function (resolve, reject) {
                                fs.stat(f, function (err, res) {
                                    if (err) {
                                        reject(err);
                                    } else {
                                        resolve(res);
                                    }
                                });
                            });
                        });
                        
                        Promise.all(promises)
                            .then(function (res) {
                                var result = res.map(function (d, i) {
                                        return {name: files[i], isDirectory: d.isDirectory()};
                                    });
                                processCallback({id: token.id, socketId: socketid, files: result, err: err}, socket);
                            }, function (err) {
                                processCallback({id: token.id, socketId: socketid, err: err}, socket);
                            });
                    } else {
                        processCallback({id: token.id, socketId: socketid, err: err}, socket);
                    }
                });
            },
            "writeDirectory": function (token, socket, socketid) {
                fs.mkdir(token.path, function (err) {
                    processCallback({id: token.id, socketId: socketid, err: err}, socket);
                });
            },
            "setMainFile": function (token, socket, socketid) {
                changeProjectSetting(token.projectName, "mainPVSFile", token.fileName)
                    .then(function (res) {
                        res.id = token.id;
                        res.socketId = socketid;
                        res.serverSent = new Date().getTime();
                        processCallback(res, socket);
                    });
            },
            "listProjects": function (token, socket, socketid) {
                var result = {id: token.id, serverSent: new Date().getTime(), socketId: socketid};
                listProjects()
                    .then(function (projects) {
                        result.projects = projects;
                        processCallback(result, socket);
                    }, function (err) {
                        result.err = err;
                        processCallback(result, socket);
                    });
            },
            "openProject": function (token, socket, socketid) {
                var res = {id: token.id, serverSent: new Date().getTime(), socketId: socketid};
                openProject(token.name)
                    .then(function (data) {
                        res.project = data;
                        processCallback(res, socket);
                    }, function (err) {
                        res.err = err;
                        processCallback(res, socket);
                    });
            },
            "createProject": function (token, socket, socketid) {
                p = pvsioProcessMap[socketid] || pvsio();
                pvsioProcessMap[socketid] = p;
                createProject(token, function (res) {
                    res.id = token.id;
                    res.socketId = socketid;
                    res.serverSent = new Date().getTime();
                    processCallback(res, socket);
                }, p);
                
            },
            "typeCheck": function (token, socket, socketid) {
                typeCheck(token.filePath, function (err, stdout, stderr) {
                    var res = {id: token.id, err: err, stdout: stdout, stderr: stderr, socketId: socketid};
                    processCallback(res, socket);
                });
            },
            "sendCommand": function (token, socket, socketid) {
                p = pvsioProcessMap[socketid];
                p.sendCommand(token.data.command, function (data) {
                    var res = {id: token.id, data: [data], socketId: socketid, type: "commandResult"};
                    processCallback(res, socket);
                });
            },
            "startProcess": function (token, socket, socketid) {
                logger.info("Calling start process for client... " + socketid);
				var root = token.data.projectName ?
                            "/public/projects/" + token.data.projectName
                            : token.data.demoName ? "/public/demos/" + token.data.demoName : "";
                p = pvsioProcessMap[socketid];
                //close the process if it exists and recreate it
                if (p) {
                    p.close('SIGTERM', true);
                    delete pvsioProcessMap[socketid];
                }
                //recreate the pvsio process
                p = pvsio();
                pvsioProcessMap[socketid] = p;
                //set the workspace dir and start the pvs process with a callback for processing process ready and exit
                //messages from the process
                p.workspaceDir(__dirname + root)
                    .start(token.data.fileName,
                        function (res) {
                            res.id = token.id;
                            res.serverSent = new Date().getTime();
                            res.socketId = socketid;
                            processCallback(res, socket);
                        });
            },
            "closeProcess": function (token, socket, socketid) {//closes pvs process
                p = pvsioProcessMap[socketid];
                var res = {id: token.id, socketId: socketid};
                if (p) {
                    p.close();
                    delete pvsioProcessMap[socketid];
                    res.message = "process closed";
                } else {
                    res.type = "attempting to close undefined process";
                }
                processCallback(res, socket);
                
            },
            "readFile": function (token, socket, socketid) {
                p = pvsioProcessMap[socketid];
                var encoding = token.encoding || "utf8";
                fs.readFile(token.filePath, encoding, function (err, content) {
                    var res = err ? {err: err} : {id: token.id, serverSent: new Date().getTime(), fileContent: content, socketId: socketid};
                    processCallback(res, socket);
                });
            },
            "writeFile": function (token, socket, socketid) {
                p = pvsioProcessMap[socketid];
                var res = {id: token.id, serverSent: new Date().getTime(), socketId: socketid};
                token.filePath = path.join(baseProjectDir, token.filePath);
                
                writeFile(token.filePath, token.fileContent, token.encoding)
                    .then(function () {
                        processCallback(res, socket);
                    }, function (err) {
                        res.err = err;
                        processCallback(res, socket);
                    });
            },
            "deleteFile": function (token, socket, socketid) {
                p = pvsioProcessMap[socketid];
                var res = {id: token.id, serverSent: new Date().getTime(), socketId: socketid};
                token.filePath = path.join(baseProjectDir, token.filePath);
                p.removeFile(token.filePath, function (err) {
                    if (!err) {
                        res.type = "fileDeleted";
                    } else {
                        res.err = err;
                    }
                    processCallback(res, socket);
                });
            }
        };

        return map;
    }

    var wsServer = new WebSocketServer({server: httpServer});
    wsServer.on("connection", function (socket) {
        var socketid =  clientid++;
        var functionMaps = createClientFunctionMaps();
        socket.on("message", function (m) {
            var token = JSON.parse(m);
            var f = functionMaps[token.type];
            if (f && typeof f === 'function') {
                //call the function with token and socket as parameter
                f(token, socket, socketid);
            } else {
                logger.warn("f is something unexpected -- I expected a function but got type " + typeof f);
            }
        });

        socket.on("close", function () {
            logger.info("closing websocket client " + socketid);
            var _p = pvsioProcessMap[socketid];
            if (_p) {
                _p.close();
            }
            delete pvsioProcessMap[socketid];
        });
    });

    wsServer.on("error", function (err) {
        logger.error(JSON.stringify(err));
    });

    httpServer.listen(port);
    logger.info("http server started .." + "now listening on port " + port);
}
run();